GPII-3843: Wait for service to be reachable before starting the Locust tests

stepanstipl opened the pull request:
This is "fix" for failing Locust tests after cluster creation. Atm. we don't wait for or check that the services are ready, but simply rely that they will be available by the time we run the Locust test.

This PR adds a wait that the targeted service needs to be reachable from within the pods before starting the test.

Depends on https://github.com/gpii-ops/docker-image-locust/pull/4

Details about my investigation can be found at https://gist.github.com/stepanstipl/6eb26d942d7ded09f741a0b9f3677de1.

TL;DR:
- Issue is related to DNS resolution
- I wasn't yet able to pin it down to specific setting, but looks like issue with negative caching somewhere in the DNS resolution path
- Issue happens only on newly-created cluster, if the tests are run right after cluster setup, otherwise it does not happen
- I suspect race-condition before all the parts of the cluster fully comes up, helped by Alpine's way of handling DNS resolution
- Tests on my personal cluster showed required number of iterations to wait for is anything between 0 and cca 30 (therefore relatively high number of retries).

One last thing I'm trying to confirm is whether this can be reproduced before the https://github.com/gpii-ops/gpii-infra/pull/342 - move to DNS challenge PR *(as I still believe it's not related to that change, but don't have good data to back that up).*

natarajaya commented on gcp/modules/locust/swarm.tf:
```diff
@@ -41,6 +41,31 @@ resource "null_resource" "locust_swarm_session" {
+        POD_READY=false
+        RETRY_COUNT=1
+        while [ "$RETRY_COUNT" -lt "$RETRIES" -a "$POD_READY" != "true" ]; do
+          CODE="$(kubectl exec -t -n locust --request-timeout=2s "$POD" -- /bin/sh -c "curl -s -o /dev/null -w '%{http_code}' --max-time 1 '${var.locust_target_host}/health'" 2>/dev/null)"
```
This will trigger pod exec alert every time it runs.

stepanstipl replied:
Thanks Sergey, I totally didn't realize this, I think I can move this inside the pod.

amatas commented on gcp/modules/locust/swarm.tf:
```diff
@@ -41,6 +41,31 @@ resource "null_resource" "locust_swarm_session" {
+        POD_READY=false
+        RETRY_COUNT=1
+        while [ "$RETRY_COUNT" -lt "$RETRIES" -a "$POD_READY" != "true" ]; do
+          CODE="$(kubectl exec -t -n locust --request-timeout=2s "$POD" -- /bin/sh -c "curl -s -o /dev/null -w '%{http_code}' --max-time 1 '${var.locust_target_host}/health'" 2>/dev/null)"
```
I guess that this will trigger the "Pod Exec Check" alert. I can not think on a better way to check this, perhaps adding this feature to the locust image as part of the application workflow.

stepanstipl replied:
Thanks Alf, yes, and I did not realize this. I think I'll move this inside the docker image.

mrtyler commented on gcp/modules/locust/swarm.tf:
```diff
@@ -41,6 +41,31 @@ resource "null_resource" "locust_swarm_session" {
         RETRY_COUNT=$(($RETRY_COUNT+1))
       done
 
+      # Wait for target service to become rechable from within the locust pod before starting the test
```
reachable

mrtyler commented on gcp/modules/locust/swarm.tf:
```diff
@@ -41,6 +41,31 @@ resource "null_resource" "locust_swarm_session" {
       done
 
+      # Wait for target service to become rechable from within the locust pod before starting the test
+      RETRIES=100
```
100 retries * 10 sec sleep = 1000 sec which 1) is kind of a weird interval (16.666 minutes) and 2) seems like kind of a long time to wait. Maybe 60 retries -> 600s/10m?

Wait, the retry counter resets for each pod, so 3000 sec/50 min in stg/prd. I think it should be shorter.

stepanstipl replied:
I think I'll need to spend some more time on this issue, generally during my tests I've seen the first pod take a lot of time, the other pods are usually ok by then, so the upper limit is really relevant for the first pod.

I've seen it needing anything from 0-30 retries, therefore quite a high number. Happy to reduce it to smth. like 60.

mrtyler commented on gcp/modules/locust/swarm.tf:
```diff
@@ -41,6 +41,31 @@ resource "null_resource" "locust_swarm_session" {
+        RETRY_COUNT=1
+        while [ "$RETRY_COUNT" -lt "$RETRIES" -a "$POD_READY" != "true" ]; do
+          CODE="$(kubectl exec -t -n locust --request-timeout=2s "$POD" -- /bin/sh -c "curl -s -o /dev/null -w '%{http_code}' --max-time 1 '${var.locust_target_host}/health'" 2>/dev/null)"
+          echo "$CODE ($RETRY_COUNT of $RETRIES)"
```
I would add some text to the output so we don't emit a number with no explanation, e.g. "Return code from $URL: $CODE ($RETRY_COUNT of $RETRIES)".

stepanstipl replied:
Ok, makes sense

mrtyler commented on gcp/modules/locust/swarm.tf:
```diff
@@ -41,6 +41,31 @@ resource "null_resource" "locust_swarm_session" {
+      RETRIES=100
+      echo "Waiting for ${var.locust_target_host} to become ready and reachable from all the locust pods..."
+      PODS=$(kubectl get pods --selector=component=worker -n locust -o=json --request-timeout=5s | jq -er '.items[].metadata.name')
+      for POD in $PODS; do
```
Hard to read the indention levels so maybe I'm missing something, but I believe that if `kubectl get pods` fails, then `PODS` will be empty, everything in this loop will be skipped, and execution will continue (and fail).

stepanstipl replied:
That is correct - my thinking here was that if the returned pods are empty (for whatever reason), the worst thing that happens is that we fail on the locust test itself. Also, it's consistent with error checking on the rest of the script, but you're right I can do more here.

stepanstipl commented:
Thanks for reviews everyone and good points. 👍 

I think that:
- a) I'll move this code inside to locust images to avoid the exec issue
- b) I need to spend a bit more time on the investigation, because it seems, however unlikely as the change doesn't touch DNS itself, that the failures are related to the move to DNS challenge.

I still think it makes sense to add some form of wait for the services to be ready.

> I wonder if locust tests are the right place for this check. In AWS, there was a deployment step that waited for preferences and flowmanager to be ready before proceeding. Maybe a general purpose wait_for step would be useful for more than just locust?
> OTOH, the locust tests replace the smoke tests that we used in AWS for this purpose, so maybe locust is the right place after all.

In this particular case the issue is that the service is up and reachable from outside, but not from within the locust pod, till the dns starts working. So test ouside would not solve the problem. 

> The first thing that springs to mind is some kind of allow list: alert on kubectl exec, unless it is kubectl exec couchdb curl /_membership or kubectl exec curl preferences/health.
> 
> Alternately, you might do something more like what Alfredo suggested and integrate this check into the worker startup scripts.

I'll try to move this inside the locust image for now - I think it's probably better aproach in this case.

stepanstipl commented:
Update:
- The loop logic has been moved into the locust image, this PR now only bumps the locust image version and depends on https://github.com/gpii-ops/docker-image-locust/pull/5
  - this is to avoid exec alerts
  - and it's actually a better place - the code is simpler
- This PR also adds dedicated cert-manager pod service account (this is to help with debugging, otherwise account is always destroyed when CI runs)

I believe I've addressed all the comments raised here, thanks 👍 !

amatas commented:
LGTM
